### ISSUE TITLE
Different gradient behaviors between `softmax_cross_entropy_with_integer_labels` and `softmax_cross_entropy`

### ISSUE THREAD

**Issue by AJINGOUP** [google-deepmind/optax#1335]
### Issue Description
I'm trying to handle dynamic batch sizes by using the `where` parameter in optax's cross-entropy functions to avoid JIT recompilation issues. While testing masked implementations, I noticed different gradient behaviors between `softmax_cross_entropy_with_integer_labels` and `softmax_cross_entropy` that seem inconsistent.

### Minimal Reproduction
```python
import jax
import jax.numpy as jnp
import optax

num_samples = 5
num_classes = 5

# Integer labels version (problematic)
def fn1(logits, labels, where):
    loss = optax.softmax_cross_entropy_with_integer_labels(
        logits=logits, 
        labels=labels, 
        where=jnp.expand_dims(where, -1)  # Expand for broadcasting
    )
    return jnp.mean(loss, where=where)

# One-hot version (works as expected)
def fn2(logits, labels, where):
    loss = optax.softmax_cross_entropy(
        logits=logits,
        labels=jax.nn.one_hot(labels, num_classes),
        where=jnp.expand_dims(where, -1)  # Same mask expansion
    )
    return jnp.mean(loss, where=where)

# Test data
logits = jax.random.normal(jax.random.PRNGKey(0), (num_samples, num_classes))
labels = jax.random.randint(jax.random.PRNGKey(1), (num_samples,), 0, num_classes)
masks = jax.random.bernoulli(jax.random.PRNGKey(2), p=0.5, shape=(num_samples,))
print("Masks:", masks)  # e.g. [False False  True False  True]

# Check fn1 (integer labels)
loss, grad = jax.value_and_grad(fn1)(logits, labels, masks)
print("fn1 grad:", grad)  # NaN in masked positions

# Check fn2 (one-hot)
loss, grad = jax.value_and_grad(fn2)(logits, labels, masks)
print("fn2 grad:", grad)  # Proper zeros in masked positions
```

### Actual Output
```
Masks: [False False  True False  True]

fn1 grad: 
[[        nan         nan         nan         nan         nan]  # ← masked
 [        nan         nan         nan         nan         nan]  # ← masked
 [-0.20860437  0.0046646   0.04716831  0.03858821  0.11818324]  # unmasked
 [        nan         nan         nan         nan         nan]  # ← masked
 [ 0.00761222  0.27513114  0.0512284  -0.38969693  0.05572514]] # unmasked

fn2 grad:
[[ 0.  0.  0.  0.  0.]  # ← properly masked
 [ 0.  0.  0.  0.  0.]  # ← properly masked
 [-0.20860437 ... ]  # unmasked
 [ 0.  0.  0.  0.  0.]  # ← properly masked
 [ 0.00761222 ... ]] # unmasked
```

### Questions
1. **Usage correctness**: Is expanding the mask via `where=jnp.expand_dims(where, -1)` the intended way to apply batch-wise masking for these functions? The docs mention broadcasting but I want to confirm this is proper usage.
2. **NaN vs zero gradients**: Why does the integer-label version produce NaNs in masked positions while the one-hot version correctly yields zeros? This causes training instability when using `softmax_cross_entropy_with_integer_labels` with masking.
3. **Dynamic batch recommendation**: What's the recommended approach for handling dynamic batch sizes with masking? The NaN gradients make `softmax_cross_entropy_with_integer_labels` unusable in this scenario despite its `where` parameter seeming designed for this purpose.

### Environment
- JAX version: 0.6.0
- Optax version: 0.2.5
- Hardware: CPU/GPU (consistent behavior)

**Would you be able to help clarify whether this is expected behavior or if I'm misunderstanding the masking implementation? Any guidance on best practices for dynamic batch sizes would be greatly appreciated. Thank you for your time and assistance!**

**Comment by rdyro** [comment]
Good catch, the gradients should be 0 for the masked examples and the behavior between the two function versions should match. This is a bug.

1. What makes sense is probably that the mask should be of shape logits or maybe 1 dimension smaller, I clarified the docstrings of both functions in PR 1335.
2. They should behave the same and 0s is a more user friendly outcome — I made both of them return 0s.
3. Typically the recommendation would be to not produce NaN gradients in the first place, with the fix you should be able to use either function!

In the future if you find any functions returning NaN gradients where you don't expect so, please let us know!

**Comment by vroulet** [comment]
Fixed by #1336
